iree/compiler/Codegen/SPIRV/SPIRVTile.cpp

```diff
@@ -0,0 +1,172 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===- SPIRVTile.cpp ------------------------------------------------------===//
+//
+// This pass tiles and Linalg ops with tensor semantics to invocations.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Codegen/Dialect/LoweringConfig.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Codegen/SPIRV/Utils.h"
+#include "iree/compiler/Codegen/Utils/MarkerUtils.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/SCF/Transforms.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-spirv-tile"
+
+namespace mlir {
+namespace iree_compiler {
+
+//===----------------------------------------------------------------------===//
+// Tiling patterns
+//===----------------------------------------------------------------------===//
+
+/// Populates `patterns` with patterns that tiles convolution/matmul ops with
+/// markers.
+static void populateTilingReductionPatterns(RewritePatternSet &patterns) {
+  MLIRContext *context = patterns.getContext();
+  auto getTileSizeFn = [&](OpBuilder &builder, Operation *op) {
+    return getTileSizes(builder, op, 2);
+  };
+  auto tilingOptions = linalg::LinalgTilingOptions()
+                           .setLoopType(linalg::LinalgTilingLoopType::Loops)
+                           .setTileSizeComputationFunction(getTileSizeFn);
+  auto marker = Identifier::get(getTileReductionMarker(), context);
+  auto filter = linalg::LinalgTransformationFilter({marker});
+
+  patterns.insert<linalg::LinalgTilingPattern<linalg::BatchMatmulOp>,
```

Comment:
Note for future. We should move to just use the op agnostic patterns here.

Reply:
Yup. This should be relatively straightforward and we can do it later.